Tutorial: Targeting feature flags on groups, pages, machines, and more (#6465)
PostHog/posthog.com: contents/tutorials/group-page-machine-flags.md

---
title: Targeting feature flags on groups, pages, machines, and more
date: 2023-08-08
author: ["ian-vanagas"]
showTitle: true
sidebar: Docs
featuredImage: ../images/tutorials/banners/tutorial-16.png
tags: ['feature flags']
---

> Most feature flags target users. To decide what value to return, PostHog’s feature flag service takes a flag key and user. What counts as a "user" or how you filter that user is up to you. Doing this enables you to target other entities like organizations, pages, machines, and more. This tutorial shows you how to target these non-user entities.

Reviewer: I found the phrase "What counts as a "user" " confusing

```suggestion
Most feature flags target users. To decide what value to return, PostHog’s feature flag service uses a flag key and an entity. Which entity to use it up to you, and these don't necessarily need to be _users_ – you can also target organizations, pages, machines, and more. 

This tutorial shows you how to target these non-user entities.
```

## Targeting groups, teams, or organizations

If you enabled [group analytics](/docs/product-analytics/group-analytics), targeting by groups, teams, or organizations is easy. When creating your feature flag, change the "Match by" value under "Release conditions" your group type name, add any conditions you want, and roll out the flag. 

![Match by organization](../images/tutorials/group-page-machine-flags/org.png)

> For this to work, you need group analytics enabled and group identification set up. Once done, you can target your flags to specific groups.

Reviewer: A bit repetitive I think
```suggestion
```

### Property or cohort filter

Another way to target an organization or team is to use the email or similar repeated identifier of a user. This enables you to filter for groups without using group analytics. Similarly, you can create a cohort for the organization, then use that as a condition.

![Email condition](../images/tutorials/group-page-machine-flags/email.png)

## Targeting pages

> Another target you might want is pages. For example, to ensure certain docs pages are consistent, beta test a new pricing page, or change a page users visited many times.

Reviewer: I think the exampels could be better here:
- I dont understand the example "certain docs pages are consistent"?
- I also didn't really understand "change a page users visited many times". It feels user based, not page based

```suggestion
Another target you might want is pages. For example, to ensure certain docs pages are consistent, test a new pricing page, or change a page users visited many times.
```

### Payloads

> You can use [feature flag payloads](/docs/feature-flags/payloads) to customize the experience depending on the page. For example, you want a button to change colors depending on the page URL for a group of users. Payloads lets you set multiple values on one feature flag which you can evaluate and use in your app.

Reviewer: Again, adding "group of users" makes me think its user based, not page based
```suggestion
You can use [feature flag payloads](/docs/feature-flags/payloads) to customize the experience depending on the page. For example, if you want a button to change colors depending on the page URL. Payloads lets you set multiple values on one feature flag which you can evaluate and use in your app.
```

To do this, create your flag with a payload such as:

```json
{
  "signup": "red",
  "login": "blue",
  "home": "green"
}
```

> In your app, get feature flag payloads, write logic to check the page URL, and use a combination of the page URL and flag payload to adjust your button. For a Next.js component, this would look like this:

Reviewer: ```suggestion
In your app, get feature flag payloads, write logic to check the page URL, and use a combination of the page URL and flag payload to adjust your button. For a react component, this would look like this:
```

> ```js

Reviewer: ```suggestion
```react
```

'use client'
import { useFeatureFlagPayload } from 'posthog-js/react'
import { usePathname } from 'next/navigation'

export default function Button() {

  const payload = useFeatureFlagPayload('button-colors')
  const pathname = usePathname()

  let color = 'blue'
  if (payload && Object.keys(payload).includes(pathname.slice(1))) {
    color = payload[pathname.slice(1)]
  } 

  return (
    <button style={{ backgroundColor: color }}>
      Click me! I am a useful button component
    </button>
  )
}
```

### Cohort

> If instead, you want to target users who visited a page repeatedly, you can create and use a static cohort. For example, if I wanted to target a flag to show a new pricing page for people who have visited the pricing page repeatedly.

Reviewer: ```suggestion
If instead, you want to target users who visited a page repeatedly, you can create and use a static cohort. For example, if you wanted to target a flag to show a new pricing page for people who have visited the pricing page repeatedly.
```

1. Create a cohort for people who had multiple pageview events
2. Duplicate as a static cohort.
3. Use the static cohort as a condition for your feature flag

> > **Why can’t I use a behavioral cohort directly?** Calculating a behavioral (based on events) cohorts is slow. Feature flags care about speed. Adding behavioral cohorts to flag evaluation would significantly slow them down. Read [a full explanation in our docs](/docs/feature-flags/common-questions#why-cant-i-use-a-cohort-with-behavioral-filters-in-my-feature-flag).

Reviewer: Nice explanation

![Cohort creation video](../images/tutorials/group-page-machine-flags/cohort.mp4)

## Targeting services, machines, devices, applications

Another entity you might want to target is services, machines, or devices. For example, you might want to roll out a [canary release](/tutorials/canary-release) to some of your servers or activate a flag in one region but not another.

### Custom user value

> PostHog does not enforce what it means to be a "user." A user can be a service, machine, device, or application and we can use feature flags to work with them this way. You do this entirely on the backend.

Reviewer: Is it important to mention "PostHog does not enforce what it means to be a "user." A user can be a service, machine, device, or application " ? Why?

Author: Rewrote this paragraph to try to explain better.

> A user for PostHog is just a distinct ID string connected to events. Although on the frontend, it is generated and connected to a real user's sessions, on the backend, it could mean anything. This includes service, machine, device, or application and we can use feature flags to work with them this way.

Reviewer: ```suggestion
A user in PostHog is just a distinct ID string connected to events. Although on the frontend, it is generated and connected to a real user's sessions, on the backend it could represent anything. This means you can use service, machine, device, or application IDs as the `distinct ID` – and feature flags will still work with them this way!
```
